windows/src/developer/kmcmpdll/CasedKeys.cpp

```diff
@@ -0,0 +1,163 @@
+
+#include "pch.h"
+
+#include <compfile.h>
+#include <compiler.h>
+#include <comperr.h>
+#include <vkeys.h>
+#include <kmcmpdll.h>
+
+#include "CharToKeyConversion.h"
+
+extern BOOL FMnemonicLayout; // TODO: these globals should be consolidated one day
+
+DWORD ExpandCapsRule(PFILE_GROUP gp, PFILE_KEY kpp, PFILE_STORE sp);
+
+PFILE_STORE FindSystemStore(PFILE_KEYBOARD fk, DWORD dwSystemID) {
+  assert(fk != NULL);
+  assert(dwSystemID != 0);
+
+  PFILE_STORE sp = fk->dpStoreArray;
+  for (DWORD i = 0; i < fk->cxStoreArray; i++, sp++) {
+    if (sp->dwSystemID == dwSystemID) {
+      return sp;
+    }
+  }
+  return NULL;
+}
+
+DWORD VerifyCasedKeys(PFILE_STORE sp) {
+  assert(sp != NULL);
+
+  if (FMnemonicLayout) {
+    // The &CasedKeys system store is not supported for
+    // mnemonic layouts in 14.0
+    return CERR_CasedKeysNotSupportedWithMnemonicLayout;
+  }
+
+  // We will rewrite this store with virtual keys
+
+  PWSTR p = sp->dpString;
+  PWSTR buf = new WCHAR[wcslen(p) * 5 + 1];  // extended keys are 5 units long, so this is the max length
+  PWSTR q = buf;
+
+  while (*p) {
+    UINT key = 0, shift = 0;
+    if (*p != UC_SENTINEL) {
+      if (!MapUSCharToVK(*p, &key, &shift)) {
+        return CERR_CasedKeysMustContainOnlyVirtualKeys;
+      }
+      if (shift & K_SHIFTFLAG) {
+        return CERR_CasedKeysMustNotIncludeShiftStates;
+      }
+    }
+    else {
+      if (*(p + 1) != CODE_EXTENDED) {
+        return CERR_CasedKeysMustContainOnlyVirtualKeys;
+      }
+      shift = *(p + 2);
+      key = *(p + 3);
+      if (shift != ISVIRTUALKEY) {
+        return CERR_CasedKeysMustNotIncludeShiftStates;
+      }
+    }
+    *q++ = UC_SENTINEL;
+    *q++ = CODE_EXTENDED;
+    *q++ = shift;
+    *q++ = key;
+    *q++ = UC_SENTINEL_EXTENDEDEND;
+    *q = 0;
+
+    p = incxstr(p);
+  }
+
+  delete[] sp->dpString;
+  sp->dpString = buf;
+
+  return CERR_None;
+}
+
+DWORD ExpandCapsRulesForGroup(PFILE_KEYBOARD fk, PFILE_GROUP gp) {
+  assert(fk != NULL);
+  assert(gp != NULL);
+
+  if (FMnemonicLayout) {
+    // The &CasedKeys system store is not supported for
+    // mnemonic layouts in 14.0
+    return CERR_None;
```

Comment:
Is this returning `CERR_NONE` so it's just skipped on mnemonic layouts?

Compared to l.35 `CERR_CasedKeysNotSupportedWithMnemonicLayout`

Reply:
Correct, we want a no-op for mnemonic layouts here.